pymongo/results.py

```diff
@@ -171,9 +171,10 @@ def upserted_id(self) -> Any:
 
     @property
     def did_upsert(self) -> bool:
-        """Whether or not an upsert took place."""
+        """Whether an upsert took place."""
         assert self.__raw_result is not None
-        return len(self.__raw_result.get("upserted", {})) > 0
+        result = self.__raw_result.get("upserted")
+        return result is not None and len(str(result)) > 0
```

Comment:
This doesn't handle the `None` case correctly.

Reply:
We need a few more tests. For both client.bulk_write and collection.update_one() that tests upserting a document with `"_id": None`.

Reply:
Do we expect something besides `False` if the result is `None`?

Reply:
yes None is a valid id